django_auth_adfs/token_manager.py

```diff
@@ -0,0 +1,489 @@
+    """
+
+    # Session key constants
+    ACCESS_TOKEN_KEY = "ADFS_ACCESS_TOKEN"
+    REFRESH_TOKEN_KEY = "ADFS_REFRESH_TOKEN"
+    TOKEN_EXPIRES_AT_KEY = "ADFS_TOKEN_EXPIRES_AT"
+    OBO_ACCESS_TOKEN_KEY = "ADFS_OBO_ACCESS_TOKEN"
+    OBO_TOKEN_EXPIRES_AT_KEY = "ADFS_OBO_TOKEN_EXPIRES_AT"
+
+    def __init__(self):
+        """Initialize the TokenManager with settings."""
+        self.refresh_threshold = getattr(settings, "TOKEN_REFRESH_THRESHOLD", 300)
+        self.store_obo_token = getattr(settings, "STORE_OBO_TOKEN", True)
+        self.logout_on_refresh_failure = getattr(
+            settings, "LOGOUT_ON_TOKEN_REFRESH_FAILURE", False
+        )
+
+        self.using_signed_cookies = (
+            django_settings.SESSION_ENGINE
+            == "django.contrib.sessions.backends.signed_cookies"
+        )
+
+        if self.using_signed_cookies:
+            logger.warning(
+                "TokenManager: Storing tokens in signed cookies is not recommended for security "
+                "reasons and cookie size limitations. Token storage will be disabled."
+            )
+
+    def store_tokens(self, request, access_token, adfs_response=None):
+        if not hasattr(request, "session"):
+            return False
+
+        try:
+            session_modified = False
+
+            encrypted_token = self.encrypt_token(access_token)
+            if encrypted_token:
+                request.session[self.ACCESS_TOKEN_KEY] = encrypted_token
+                session_modified = True
+                logger.debug("Stored access token")
+
+            if adfs_response and "refresh_token" in adfs_response:
+                refresh_token = adfs_response["refresh_token"]
+                if refresh_token:
+                    encrypted_token = self.encrypt_token(refresh_token)
+                    if encrypted_token:
+                        request.session[self.REFRESH_TOKEN_KEY] = encrypted_token
+                        session_modified = True
+                        logger.debug("Stored refresh token")
+                    else:
+                        logger.warning("Failed to encrypt refresh token")
+                else:
+                    logger.warning("Empty refresh token received from ADFS")
+            else:
+                logger.debug("No refresh token in ADFS response")
+
+            if adfs_response and "expires_in" in adfs_response:
+                expires_at = datetime.datetime.now() + datetime.timedelta(
+                    seconds=int(adfs_response["expires_in"])
+                )
+                request.session[self.TOKEN_EXPIRES_AT_KEY] = expires_at.isoformat()
+                session_modified = True
+                logger.debug("Stored token expiration")
+
+            if self.store_obo_token:
+                try:
+                    from django_auth_adfs.backend import AdfsBaseBackend
+
+                    backend = AdfsBaseBackend()
+                    obo_token = backend.get_obo_access_token(access_token)
```

Comment:
This implies there's a circular dependency in the project. Can we refactor that out? Or did you have other thoughts on it?